### PR TITLE
Fix #2603: Default values for non-exposed defs.

### LINKED ISSUE #2603: Inner def with default param in a Scala.js-defined JS class produces invalid IR
Reproduction:
```scala
import scala.scalajs.js
import js.annotation._

@ScalaJSDefined
class Foo extends js.Object {
  def bar(): Int = {
    def inner(start: Int = 0): Int = start + 1
    inner()
  }
}

object HelloWorld extends js.JSApp {
  def main() {
    val foo = new Foo
    assert(foo.bar() == 1)
  }
}
```
IR checker error:
```
[error] file:/localhome/doeraene/projects/scalajs/examples/helloworld/HelloWorld.scala(15:5:UndefinedParam): Invalid expression tree
```

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by gzm0** [comment]
Review by @sjrd

File: compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala [hunk 1]
```diff
@@ -1998,7 +1998,24 @@ abstract class GenJSCode extends plugins.PluginComponent
           isRawJSCtorDefaultParam(sym)
         } else {
           sym.hasFlag(reflect.internal.Flags.DEFAULTPARAM) &&
-          isRawJSType(sym.owner.tpe)
+          isRawJSType(sym.owner.tpe) && {
+            /* If this is a default parameter accessor on a
+             * ScalaJSDefinedJSClass, we need to know if the method for which we
+             * are the default parameter is exposed or not.
+             * We do this by looking up the de-mangled name in the owner. Note
```
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala]: "de-mangled" makes me think about `unencodedName` much more than what's happening here. Consider something like "We do this by removing the `$default` suffix from the method name, and looking up a member with that name in the owner."

File: compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala [hunk 2]
```diff
@@ -1998,7 +1998,24 @@ abstract class GenJSCode extends plugins.PluginComponent
...
+             * that this does not work for local methods. But local methods are
+             * never exposed.
+             * Further note that overloads are easy, because either all or none
+             * of them are exposed.
+             */
+            def isOwnerMethodExposed = {
```
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala]: "owner" is not very accurate wrt what scalac calls an owner. Consider maybe `isAttachedMethodExposed`?

File: test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/ScalaJSDefinedTest.scala [hunk 3]
```diff
@@ -1307,6 +1307,19 @@ class ScalaJSDefinedTest {
     assertEquals("hello3 4", dyn.foo("hello", 4))
   }
 
+  @Test def overload_with_default_parameter(): Unit = {
+    @ScalaJSDefined
+    class OverloadDefaultParameter extends js.Object {
+      def foo(x: Int) = x
+      def foo(x: String = "") = x
```
Comment (sjrd) [inline test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/ScalaJSDefinedTest.scala]: Public methods need result type (I think Scalastyle will complain about this).

**Comment by sjrd** [comment]
What made the issues you were describing in #2603 go away?

**Comment by gzm0** [comment]
> What made the issues you were describing in #2603 go away?

Checking for the existence of the `Symbol` before feeding it to `isExposed`.

**Comment by gzm0** [comment]
Updated.

**Pull request merged**